src/Workspaces/Core/Portable/Workspace/Host/Mef/ExportWorkspaceServiceFactoryAttribute.cs

```diff
@@ -2,45 +2,46 @@
 // The .NET Foundation licenses this file to you under the MIT license.
 // See the LICENSE file in the project root for more information.
 
-#nullable disable
-
 using System;
+using System.Collections.Generic;
+using System.Collections.Immutable;
 using System.Composition;
 
 namespace Microsoft.CodeAnalysis.Host.Mef
 {
     /// <summary>
     /// Use this attribute to declare a <see cref="IWorkspaceServiceFactory"/> implementation for inclusion in a MEF-based workspace.
     /// </summary>
+    /// <remarks>
+    /// Declares a <see cref="IWorkspaceServiceFactory"/> implementation for inclusion in a MEF-based workspace.
+    /// </remarks>
+    /// <param name="serviceType">The type that will be used to retrieve the service from a <see cref="HostWorkspaceServices"/>.</param>
+    /// <param name="layer">The layer or workspace kind that the service is specified for; <see cref="ServiceLayer.Default" />, <see cref="WorkspaceKind.MiscellaneousFiles" />etc.</param>
     [MetadataAttribute]
     [AttributeUsage(AttributeTargets.Class)]
-    public class ExportWorkspaceServiceFactoryAttribute : ExportAttribute
+    public class ExportWorkspaceServiceFactoryAttribute(Type serviceType, string layer = ServiceLayer.Default) : ExportAttribute(typeof(IWorkspaceServiceFactory))
     {
         /// <summary>
         /// The assembly qualified name of the service's type.
         /// </summary>
-        public string ServiceType { get; }
+        public string ServiceType { get; } = LayeredServiceUtilities.GetAssemblyQualifiedServiceTypeName(serviceType, nameof(serviceType));
 
         /// <summary>
         /// The layer that the service is specified for; ServiceLayer.Default, etc.
         /// </summary>
-        public string Layer { get; }
+        public string Layer { get; } = layer ?? throw new ArgumentNullException(nameof(layer));
 
         /// <summary>
-        /// Declares a <see cref="IWorkspaceServiceFactory"/> implementation for inclusion in a MEF-based workspace.
+        /// <see cref="WorkspaceKind"/>s that the service is specified for.
+        /// If non-empty the service is only exported for the listed workspace kinds and <see cref="Layer"/> is not applied,
+        /// unless <see cref="Layer"/> is is ignored if non-empty.
```

Comment:
part after the comma isn't clear to me.

Reply:
Oh, will rephrase - it didn't save (VS crashed).